File: web/src/features/channels/lib/coding-plan.ts
/*
Copyright (C) 2023-2026 QuantumNous

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.

For commercial licensing, please contact support@quantumnous.com
*/
import type { AdvancedCustomConfig, AdvancedCustomRoute } from '../types'

export const CHANNEL_TYPE_CODING_PLAN = 59

export type CodingPlanProvider = {
  value: string
  label: string
  openAIBaseUrl: string
  anthropicBaseUrl: string
  responsesBaseUrl?: string
  modelListBaseUrl?: string
}

export const CODING_PLAN_PROVIDER_OPTIONS: CodingPlanProvider[] = [
  {
    value: 'glm-coding-plan',
    label: 'GLM Coding Plan',
    openAIBaseUrl: 'https://open.bigmodel.cn/api/coding/paas/v4',
    anthropicBaseUrl: 'https://open.bigmodel.cn/api/anthropic',
    modelListBaseUrl: 'https://open.bigmodel.cn/api/coding/paas/v4',
  },
  {
    value: 'glm-coding-plan-international',
    label: 'Z.ai Coding Plan',
    openAIBaseUrl: 'https://api.z.ai/api/coding/paas/v4',
    anthropicBaseUrl: 'https://api.z.ai/api/anthropic',
    modelListBaseUrl: 'https://api.z.ai/api/coding/paas/v4',
  },
  {
    value: 'kimi-coding-plan',
    label: 'Kimi Code',
    openAIBaseUrl: 'https://api.kimi.com/coding/v1',
    anthropicBaseUrl: 'https://api.kimi.com/coding',
    responsesBaseUrl: 'https://api.kimi.com/coding/v1',
    modelListBaseUrl: 'https://api.kimi.com/coding/v1',
  },
  {
    value: 'doubao-coding-plan',
    label: 'Doubao Coding Plan',
    openAIBaseUrl: 'https://ark.cn-beijing.volces.com/api/coding/v3',
    anthropicBaseUrl: 'https://ark.cn-beijing.volces.com/api/coding',
    responsesBaseUrl: 'https://ark.cn-beijing.volces.com/api/coding/v3',
    modelListBaseUrl: 'https://ark.cn-beijing.volces.com/api/coding/v3',
  },
  {
    value: 'qwen-coding-plan',
    label: 'Qwen Coding Plan',
    openAIBaseUrl: 'https://coding-intl.dashscope.aliyuncs.com/v1',
    anthropicBaseUrl:
      'https://coding-intl.dashscope.aliyuncs.com/apps/anthropic',
    modelListBaseUrl: 'https://coding-intl.dashscope.aliyuncs.com/v1',
  },
  {
    value: 'qwen-token-plan',
    label: 'Qwen Token Plan',
    openAIBaseUrl:
      'https://token-plan.ap-southeast-1.maas.aliyuncs.com/compatible-mode/v1',
    anthropicBaseUrl:
      'https://token-plan.ap-southeast-1.maas.aliyuncs.com/apps/anthropic',
    responsesBaseUrl:
      'https://token-plan.ap-southeast-1.maas.aliyuncs.com/compatible-mode/v1',
    modelListBaseUrl:
      'https://token-plan.ap-southeast-1.maas.aliyuncs.com/compatible-mode/v1',
  },
  {
    value: 'minimax-token-plan',
    label: 'MiniMax Token Plan',
    openAIBaseUrl: 'https://api.minimax.io/v1',
    anthropicBaseUrl: 'https://api.minimax.io/anthropic',
    responsesBaseUrl: 'https://api.minimax.io/v1',
    modelListBaseUrl: 'https://api.minimax.io/v1',
  },
  {
    value: 'opencode-go',
    label: 'OpenCode Go',
    openAIBaseUrl: 'https://opencode.ai/zen/go/v1',
    anthropicBaseUrl: 'https://opencode.ai/zen/go/v1',
    modelListBaseUrl: 'https://opencode.ai/zen/go/v1',
  },
]

export function getDefaultCodingPlanProvider(): string {
  return CODING_PLAN_PROVIDER_OPTIONS[0]?.value || ''
}

export function normalizeCodingPlanProvider(value: string | undefined): string {
  const normalized = String(value || '')
    .trim()
    .replace(/\/+$/, '')
  const provider = CODING_PLAN_PROVIDER_OPTIONS.find(
    (option) =>
      option.value === normalized ||
      option.openAIBaseUrl === normalized ||
      option.anthropicBaseUrl === normalized ||
      option.responsesBaseUrl === normalized
  )
  return provider?.value || normalized
}

export function buildCodingPlanAdvancedCustomConfig(
  providerValue: string | undefined
): AdvancedCustomConfig | null {
  const provider = CODING_PLAN_PROVIDER_OPTIONS.find(
    (option) => option.value === normalizeCodingPlanProvider(providerValue)
  )
  if (!provider) return null

  const routes: AdvancedCustomRoute[] = []
  if (provider.openAIBaseUrl) {
    const chatUrl = joinProviderPath(provider.openAIBaseUrl, 'chat/completions')
    routes.push(
      codingPlanRoute('/v1/chat/completions', chatUrl, 'none'),
      codingPlanRoute(
        '/v1/completions',
        joinProviderPath(provider.openAIBaseUrl, 'completions'),
        'none'
      ),
      codingPlanRoute(
        '/v1beta/models/{model}:generateContent',
        chatUrl,
        'gemini_generate_content_to_openai_chat_completions'
      )
    )
    if (!provider.responsesBaseUrl) {
      routes.push(
        codingPlanRoute(
          '/v1/responses',
          chatUrl,
          'openai_responses_to_openai_chat_completions'
        )
      )
    }
  }
  if (provider.responsesBaseUrl) {
    routes.push(
      codingPlanRoute(
        '/v1/responses',
        joinProviderPath(provider.responsesBaseUrl, 'responses'),
        'none'
      )
    )
  }
  if (provider.anthropicBaseUrl) {
    routes.push(
      codingPlanRoute(
        '/v1/messages',
        joinProviderPath(provider.anthropicBaseUrl, 'v1/messages'),
        'none'
      )
    )
  }
  if (provider.modelListBaseUrl) {
    routes.push(
      codingPlanRoute(
        '/v1/models',
        joinProviderPath(provider.modelListBaseUrl, 'models'),
        'none'
      )
    )
  }

  return { advanced_routes: routes }
}

function codingPlanRoute(
  incomingPath: string,
  upstreamPath: string,
  converter: AdvancedCustomRoute['converter']
): AdvancedCustomRoute {
  return {
    incoming_path: incomingPath,
    upstream_path: upstreamPath,
    converter,
    auth: {
      type: 'header',
      name: 'Authorization',
      value: 'Bearer {api_key}',
    },
  }
}

function joinProviderPath(baseUrl: string, path: string): string {
  return `${baseUrl.trim().replace(/\/+$/, '')}/${path.trim().replace(/^\/+/, '')}`
}

File: relaykit/dto/coding_plan.go
package dto

import (
	"fmt"
	"strings"
)

const (
	CodingPlanProviderGLMChina      = "glm-coding-plan"
	CodingPlanProviderGLMGlobal     = "glm-coding-plan-international"
	CodingPlanProviderKimi          = "kimi-coding-plan"
	CodingPlanProviderDoubao        = "doubao-coding-plan"
	CodingPlanProviderQwenCoding    = "qwen-coding-plan"
	CodingPlanProviderQwenTokenPlan = "qwen-token-plan"
	CodingPlanProviderMiniMax       = "minimax-token-plan"
	CodingPlanProviderOpenCodeGo    = "opencode-go"
	CodingPlanProviderCustom        = "custom-coding-plan"
)

type CodingPlanPreset struct {
	ID               string
	OpenAIBaseURL    string
	AnthropicBaseURL string
	ResponsesBaseURL string
	ModelListBaseURL string
}

var codingPlanPresets = []CodingPlanPreset{
	{ID: CodingPlanProviderGLMChina, OpenAIBaseURL: "https://open.bigmodel.cn/api/coding/paas/v4", AnthropicBaseURL: "https://open.bigmodel.cn/api/anthropic", ModelListBaseURL: "https://open.bigmodel.cn/api/coding/paas/v4"},
	{ID: CodingPlanProviderGLMGlobal, OpenAIBaseURL: "https://api.z.ai/api/coding/paas/v4", AnthropicBaseURL: "https://api.z.ai/api/anthropic", ModelListBaseURL: "https://api.z.ai/api/coding/paas/v4"},
	{ID: CodingPlanProviderKimi, OpenAIBaseURL: "https://api.kimi.com/coding/v1", AnthropicBaseURL: "https://api.kimi.com/coding", ResponsesBaseURL: "https://api.kimi.com/coding/v1", ModelListBaseURL: "https://api.kimi.com/coding/v1"},
	{ID: CodingPlanProviderDoubao, OpenAIBaseURL: "https://ark.cn-beijing.volces.com/api/coding/v3", AnthropicBaseURL: "https://ark.cn-beijing.volces.com/api/coding", ResponsesBaseURL: "https://ark.cn-beijing.volces.com/api/coding/v3", ModelListBaseURL: "https://ark.cn-beijing.volces.com/api/coding/v3"},
	{ID: CodingPlanProviderQwenCoding, OpenAIBaseURL: "https://coding-intl.dashscope.aliyuncs.com/v1", AnthropicBaseURL: "https://coding-intl.dashscope.aliyuncs.com/apps/anthropic", ModelListBaseURL: "https://coding-intl.dashscope.aliyuncs.com/v1"},
	{ID: CodingPlanProviderQwenTokenPlan, OpenAIBaseURL: "https://token-plan.ap-southeast-1.maas.aliyuncs.com/compatible-mode/v1", AnthropicBaseURL: "https://token-plan.ap-southeast-1.maas.aliyuncs.com/apps/anthropic", ResponsesBaseURL: "https://token-plan.ap-southeast-1.maas.aliyuncs.com/compatible-mode/v1", ModelListBaseURL: "https://token-plan.ap-southeast-1.maas.aliyuncs.com/compatible-mode/v1"},
	{ID: CodingPlanProviderMiniMax, OpenAIBaseURL: "https://api.minimax.io/v1", AnthropicBaseURL: "https://api.minimax.io/anthropic", ResponsesBaseURL: "https://api.minimax.io/v1", ModelListBaseURL: "https://api.minimax.io/v1"},
	{ID: CodingPlanProviderOpenCodeGo, OpenAIBaseURL: "https://opencode.ai/zen/go/v1", AnthropicBaseURL: "https://opencode.ai/zen/go", ModelListBaseURL: "https://opencode.ai/zen/go/v1"},
	{ID: CodingPlanProviderCustom, OpenAIBaseURL: "https://your-openai-compatible-base-url.example/v1", AnthropicBaseURL: "https://your-anthropic-compatible-base-url.example", ResponsesBaseURL: "https://your-openai-compatible-base-url.example/v1", ModelListBaseURL: "https://your-openai-compatible-base-url.example/v1"},
}

func ResolveCodingPlanPreset(value string) (CodingPlanPreset, bool) {
	value = normalizeCodingPlanPresetValue(value)
	if value == "" {
		return CodingPlanPreset{}, false
	}
	for _, preset := range codingPlanPresets {
		if strings.EqualFold(value, preset.ID) || strings.EqualFold(value, normalizeCodingPlanPresetValue(preset.OpenAIBaseURL)) || strings.EqualFold(value, normalizeCodingPlanPresetValue(preset.AnthropicBaseURL)) || strings.EqualFold(value, normalizeCodingPlanPresetValue(preset.ResponsesBaseURL)) {
			return preset, true
		}
	}
	return CodingPlanPreset{}, false
}

func (s *ChannelOtherSettings) ApplyCodingPlanPreset(provider string) error {
	if s == nil {
		return fmt.Errorf("coding plan settings are required")
	}
	preset, ok := ResolveCodingPlanPreset(provider)
	if !ok {
		return fmt.Errorf("unknown coding plan provider: %s", strings.TrimSpace(provider))
	}
	config := preset.AdvancedCustomConfig()
	if err := config.Validate(); err != nil {
		return err
	}
	s.CodingPlanProvider = preset.ID
	s.AdvancedCustom = config
	return nil
}

func (p CodingPlanPreset) AdvancedCustomConfig() *AdvancedCustomConfig {
	routes := make([]AdvancedCustomRoute, 0, 7)
	if p.OpenAIBaseURL != "" {
		chatURL := codingPlanJoinPath(p.OpenAIBaseURL, "chat/completions")
		imageURL := codingPlanJoinPath(p.OpenAIBaseURL, "images/generations")
		routes = append(routes, codingPlanRoute(advancedCustomEndpointPathOpenAIChat, chatURL, advancedCustomConverterNone), codingPlanRoute(advancedCustomEndpointPathOpenAICompletions, chatURL, advancedCustomConverterOpenAICompletionsToChat), codingPlanRoute("/v1beta/models/{model}:generateContent", chatURL, advancedCustomConverterGeminiContentToOpenAIChat), codingPlanRoute(advancedCustomEndpointPathImageGeneration, imageURL, advancedCustomConverterNone))
		if p.ID == CodingPlanProviderOpenCodeGo {
			responsesRoute := codingPlanRoute(advancedCustomEndpointPathOpenAIResponses, codingPlanJoinPath(p.OpenAIBaseURL, "responses"), advancedCustomConverterNone)
			responsesRoute.Models = []string{"deepseek-v4-flash"}
			routes = append(routes, responsesRoute)
		}
		if p.ResponsesBaseURL == "" {
			routes = append(routes, codingPlanRoute(advancedCustomEndpointPathOpenAIResponses, chatURL, advancedCustomConverterOpenAIResponsesToOpenAIChat))
		}
	}
	if p.ResponsesBaseURL != "" {
		routes = append(routes, codingPlanRoute(advancedCustomEndpointPathOpenAIResponses, codingPlanJoinPath(p.ResponsesBaseURL, "responses"), advancedCustomConverterNone))
	}
	if p.AnthropicBaseURL != "" {
		routes = append(routes, codingPlanRoute(advancedCustomEndpointPathClaudeMessages, codingPlanJoinPath(p.AnthropicBaseURL, "v1/messages"), advancedCustomConverterNone))
	}
	if p.ModelListBaseURL != "" {
		routes = append(routes, codingPlanRoute(AdvancedCustomModelListPath, codingPlanJoinPath(p.ModelListBaseURL, "models"), advancedCustomConverterNone))
	}
	return &AdvancedCustomConfig{Routes: routes}
}

func codingPlanRoute(incomingPath, upstreamPath, converter string) AdvancedCustomRoute {
	return AdvancedCustomRoute{IncomingPath: incomingPath, UpstreamPath: upstreamPath, Converter: converter, Auth: &AdvancedCustomRouteAuth{Type: AdvancedCustomAuthTypeHeader, Name: "Authorization", Value: "Bearer {api_key}"}}
}

func codingPlanJoinPath(baseURL, path string) string {
	return strings.TrimRight(strings.TrimSpace(baseURL), "/") + "/" + strings.TrimLeft(strings.TrimSpace(path), "/")
}

func normalizeCodingPlanPresetValue(value string) string {
	return strings.TrimRight(strings.TrimSpace(value), "/")
}
